worldedit-core/src/main/java/com/sk89q/worldedit/internal/util/RecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,234 @@
+     */
+    public record DirectoryDeletedEvent(Path path) implements DirEntryChangeEvent {
+    }
+
+
+    private static final Logger LOGGER = LogManagerCompat.getLogger();
+
+    private final Path root;
+    private final WatchService watchService;
+    private Thread watchThread;
+    private Consumer<DirEntryChangeEvent> eventConsumer;
+    private BiMap<WatchKey, Path> watchRootMap = HashBiMap.create();
+
+    private RecursiveDirectoryWatcher(Path root, WatchService watchService) {
+        this.root = root;
+        this.watchService = watchService;
+    }
+
+    /**
+     * Create a new recursive directory watcher for the given root folder.
+     * You have to call {@link #start(Consumer)} before the instance starts monitoring.
+     *
+     * @param root Folder to watch for changed files recursively.
+     * @return a new instance that will monitor the given root folder
+     * @throws IOException If creating the watcher failed, e.g. due to root not existing.
+     */
+    public static Optional<RecursiveDirectoryWatcher> create(Path root) throws IOException {
+        try {
+            WatchService watchService = root.getFileSystem().newWatchService();
+            return Optional.of(new RecursiveDirectoryWatcher(root, watchService));
+        } catch (UnsupportedOperationException ignored) {
+            return Optional.empty();
+        }
+    }
+
+    private void registerFolderWatcher(Path root) throws IOException {
+        WatchKey watchKey = root.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE, StandardWatchEventKinds.ENTRY_MODIFY);
+        LOGGER.debug("Watch registered: " + root);
+        watchRootMap.put(watchKey, root);
+    }
+
+    private void triggerInitialEvents(Path root) throws IOException, FilenameException {
+        Path schematicRoot = WorldEdit.getInstance().getSchematicsManager().getRoot();
+        eventConsumer.accept(new DirectoryCreatedEvent(root));
+        for (Path path : Files.newDirectoryStream(root)) {
+            path = WorldEdit.getInstance().getSafeOpenFile(null, schematicRoot.toFile(), schematicRoot.relativize(path).toString(), null).toPath();
+            if (Files.isDirectory(path)) {
+                triggerInitialEvents(path);
+            } else {
+                eventConsumer.accept(new FileCreatedEvent(path));
+            }
+        }
+    }
+
+    /**
+     * Make this RecursiveDirectoryWatcher instance start monitoring the root folder it was created on.
+     * When this is called, RecursiveDirectoryWatcher will send initial notifications for the entire
+     * file structure in the configured root.
+     * @param eventConsumer The lambda that's fired for every file event.
+     */
+    public void start(Consumer<DirEntryChangeEvent> eventConsumer) {
+        Path schematicRoot = WorldEdit.getInstance().getSchematicsManager().getRoot();
+
+        this.eventConsumer = eventConsumer;
+        watchThread = new Thread(() -> {
```

Comment:
Just in case, we should check for an existing `watchThread` and kill it or throw. Killing requires us to have a means of communicating to it.